web/source/keyboards/kmwkeyboards.ts

```diff
@@ -893,30 +887,51 @@ namespace com.keyman.keyboards {
       }
     };
 
+    /**
+     * Return a Promise of the merged keyboard stubs and error stubs
+     * @param {(KeyboardStub|ErrorStub)[]} keyboardStubs 
+     * @param {ErrorStub[]} errorStubs 
+     * @returns {Promise<KeyboardStub|ErrorStub)[]>}
+     */
+    private mergeStubs(keyboardStubs: (KeyboardStub|ErrorStub)[], errorStubs: ErrorStub[]) :
+        Promise<(KeyboardStub|ErrorStub)[]> {
+      if (errorStubs.length == 0) {
+        return Promise.resolve(keyboardStubs);
+      } if (keyboardStubs.length == 0) {
+        return Promise.reject(errorStubs);
+      } else {
+        // Merge this with errorStubs
+        let result: (KeyboardStub|ErrorStub)[] = keyboardStubs;
+        return Promise.resolve(result.concat(errorStubs));
+      }
+    }
+
     /**
      * Build 362: addKeyboardArray() link to Cloud. One or more arguments may be used
      *
-     * @param {string|Object} x keyboard name string or keyboard metadata JSON object
+     * @param {(string|KeyboardStub)[]} x keyboard name string or keyboard metadata JSON object
      *
      */
-    addKeyboardArray(x: any[] | IArguments): void {
-      // Store all keyboard meta-data for registering later if called before initialization
+    async addKeyboardArray(x: (string|KeyboardStub)[]): Promise<(KeyboardStub|ErrorStub)[]> {
```

Comment:
Are we sure we want to use `async` within our library code? It might be better to always use promise style, to avoid pollution of `async`/`await` tagging up the function chains.

Reply:
I found it necessary to handle adding keyboards before the language list or keyman web is initialized.

e.g. the previous uses of  `this.deferredStubs`

Reply:
You still don't need to use `async`/`await` to accomplish this. You can wrap it in a traditional promise pattern instead. It means restructuring the internals of the function.

Reply:
I *think* from our discussion in standup, the async/await calls should be OK?
This doesn't affect KMW consumers, since the add*Keyboard() functions previously weren't returning Promises.
Now they are, and I just mark them "async" so the internal await can be guaranteed.

A TODO chore would be to have some consistency in the other internal KMW promises.

Reply:
I _think_ both approaches are valid.

> A TODO chore would be to have some consistency in the other internal KMW promises.

One could just as easily argue that it would have been nice for this PR set to maintain consistency with the existing pattern, which would be what @mcdurdin suggested above.  While I'm not stating it's required... the fact that `async/await` is the new pattern lends it less weight unless we can show the benefits outweigh the costs.

To me, this bit from earlier:

> Are we sure we want to use async within our library code? It might be better to always use promise style, to avoid pollution of async/await tagging up the function chains.

implies concern about code debugging.  Both `Promises` and `await/async` are handled... interestingly in debugging call stacks, and we probably would prefer to use whichever pattern gives us the better call stack information for debugging.  If using `await/async` costs us maintainability, that'd be an issue.